java-spiffe-core/src/main/java/spiffe/workloadapi/Address.java

```diff
@@ -0,0 +1,156 @@
+package spiffe.workloadapi;
+
+import lombok.val;
+import org.apache.commons.lang3.StringUtils;
+import spiffe.exception.SocketEndpointAddressException;
+
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Utility class to get the default Workload API address and parse string addresses.
+ */
+public class Address {
+
+    /**
+     * 	Environment variable holding the default Workload API address.
+     */
+    public static final String SOCKET_ENV_VARIABLE = "SPIFFE_ENDPOINT_SOCKET";
+
+    private static final List<String> VALID_SCHEMES = Arrays.asList("unix", "tcp");
+
+    /**
+     * Returns the default Workload API address hold by the system environment variable
+     * defined by SOCKET_ENV_VARIABLE
+     */
+    public static String getDefaultAddress() {
+        return System.getenv(Address.SOCKET_ENV_VARIABLE);
+    }
+
+    /**
+     * Parses and validates a Workload API socket address.
+     * <p>
+     * The given address should either have a tcp, or a unix scheme.
+     * <p>
+     * The given address should contain a path.
+     * <p>
+     * The given address cannot be opaque, cannot have fragments, query values or user info.
+     * <p>
+     * If the given address is tcp, it should contain an IP and a port.
+     *
+     * @param address the Workload API socket address as a string
+     * @return an instance of a {@link URI}
+     *
+     * @throws SocketEndpointAddressException if the address could not be parsed or if it is not valid
+     */
+    public static URI parseAddress(String address) throws SocketEndpointAddressException {
+        URI parsedAddress;
+        try {
+            parsedAddress = new URI(address);
+        } catch (URISyntaxException e) {
+            throw new SocketEndpointAddressException(String.format("Workload endpoint socket is not a valid URI: %s", address), e);
+        }
+
+        val scheme = parsedAddress.getScheme();
+        if (!isValid(scheme)) {
+            throw new SocketEndpointAddressException(String.format("Workload endpoint socket URI must have a tcp:// or unix:// scheme: %s", address));
+        }
+
+        String error = null;
+        switch (scheme) {
+            case "unix":
+                error = validateUnixAddress(parsedAddress);
+                break;
+            case "tcp":
+                error = validateTcpAddress(parsedAddress);
+                break;
+            default:
+                error = "Workload endpoint socket URI must have a tcp:// or unix:// scheme: %s";
+        }
+
+        if (StringUtils.isNotBlank(error)) {
+            throw new SocketEndpointAddressException(String.format(error, address));
+        }
```

Comment:
Would it make sense for you if the validateUnixAddress() and validateTcpAddress() methods are the ones that throw the exception instead of returning the error string? This way you could delete the `if (StringUtils.isNotBlank(error))`, but still have to throw the SocketEndpointAddressException for the default case.

Reply:
Tried this suggestion but the code ends up being a lot more complicated, with that `if isNotBlank` duplicated in the validate methods and complicated `if/else` structures in those methods.

Reply:
I was thinking more in something along these lines:
```java
        switch (scheme) {
            case "unix":
            	validateUnixAddress(parsedAddress);
                break;
            case "tcp":
            	validateTcpAddress(parsedAddress);
                break;
            default:
            	throw new SocketEndpointAddressException(String.format("Workload endpoint socket URI must have a tcp:// or unix:// scheme: %s", address));
        }

        return parsedAddress;
    }

    private static void validateUnixAddress(URI parsedAddress) throws SocketEndpointAddressException {
        if (parsedAddress.isOpaque() && parsedAddress.isAbsolute()) {
        	throw new SocketEndpointAddressException(String.format("Workload endpoint unix socket URI must not be opaque: %s", parsedAddress));
        }

        if (StringUtils.isNotBlank(parsedAddress.getUserInfo())) {
        	throw new SocketEndpointAddressException(String.format("Workload endpoint unix socket URI must not include user info: %s", parsedAddress));
        }

        if (StringUtils.isBlank(parsedAddress.getHost()) && StringUtils.isBlank(parsedAddress.getPath())) {
        	throw new SocketEndpointAddressException(String.format("Workload endpoint unix socket URI must include a path: %s", parsedAddress));
        }

        if (StringUtils.isNotBlank(parsedAddress.getRawQuery())) {
        	throw new SocketEndpointAddressException(String.format("Workload endpoint unix socket URI must not include query values: %s", parsedAddress));
        }

        if (StringUtils.isNotBlank(parsedAddress.getFragment())) {
        	throw new SocketEndpointAddressException(String.format("Workload endpoint unix socket URI must not include a fragment: %s", parsedAddress));
        }
    }
```
I agree it ends up being more verbose, but I believe that in terms of logic it is simpler.
Anyways, feel free to dismiss this comment, no strong feelings for this.

Reply:
yeah, I thought about that option as well, but not sure I like it either